src/AssetsListener.php

```diff
@@ -0,0 +1,367 @@
+        ]);
+        return $response;
+    }
+
+    protected function filter(MvcEvent $event, $assetFile, $assetItem)
+    {
+        $filters = $this->getAssetsManager()->getAssetFilters($assetItem);
+        if (!$filters) {
+            return fopen($assetFile, 'r');
+        }
+
+        $assetContent = file_get_contents($assetFile);
+        foreach($filters as $filter) {
+            if (!$filter instanceof FilterInterface) {
+                $filter = $this->getFilterManager()->get($filter);
+            }
+            $assetContent = $filter->filter($assetContent);
+        }
+        return $assetContent;
+    }
+
+    protected function cache(MvcEvent $event, $source)
+    {
+        if (!$this->isCacheToPublic) {
+            return $source;
+        }
+
+        $request = $event->getRequest();
+        $targetFile = $this->getAssetsManager()->getPublicFolder() . substr($request->getRequestUri(), strlen($request->getBasePath()));
+
+        $targetDir  = dirname($targetFile);
+        if (!file_exists($targetDir) && @mkdir($targetDir, 0777, true) === false) {
+            throw new Exception\RuntimeException('can not create folder for caching asset');
+        }
+
+        if (is_string($source)) {
+            if (@file_put_contents($targetFile, $source) !== false) {
+                return $source;
+            }
+            throw new Exception\RuntimeException('can not save file to cache');
+        }
+
+        if (is_resource($source)) {
+            $target = fopen($targetFile, 'x+');
+            if (@stream_copy_to_stream($source, $target)) {
+                fclose($source);
+                fseek($target, 0);
+                return $target;
+            }
+            throw new Exception\RuntimeException('can not save file to cache');
+        }
+
+        throw new Exception\InvalidArgumentException(sprintf(
+            '%s: expects "$source" parameter an string or resource, received "%s"',
+            __METHOD__,
+            (is_object($source) ? get_class($source) : gettype($source))
+        ));
+    }
+
+    public function setCacheToPublic($flag)
+    {
+        $this->isCacheToPublic = (bool)$flag;
+        return $this;
+    }
+
+    public function isCacheToPublic()
+    {
+        return $this->isCacheToPublic;
+    }
+
+    /**
+     * @param string $routeName
+     * @return self
+     */
+    public function setRouteName($routeName)
+    {
+        if (!is_string($routeName)) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s: expects parameter an %s, received "%s"',
+                __METHOD__,
+                'string',
+                (is_object($routeName) ? get_class($routeName) : gettype($routeName))
+            ));
+        }
+        $this->routeName = $routeName;
+        return $this;
+    }
+
+    /**
+     * @param array|RouteStackInterface $router
+     * @return self
+     */
+    public function setRouter($router)
```

Comment:
In what cases would an `array` be a suitable router? Shouldn't just requiring an interface be a requirement?